### PR TITLE
Scf/assignment8

### PR DESCRIPTION
**Pull request by meta-Niraj-Gupta**

### CODE REVIEW

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Animals.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		this.weight = weight;
+		this.age = age;
+		this.sound = sound;
+		this.numberOfLegs = numberOfLegs;
+		this.category = category;
+		this.animalType = animalType;
+	}
+
+	/**
+	 * @return : The weight of the animal.
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Animals.java]: include method description in all getters

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Lion.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package ZooManagementSystem;
+
+/**
+ * @author Niraj Gupta
+ * Class to represent the Lion which is also the sub class of the class Mammals and Mammals is the sub class of the class Animal.
+ *
+ */
+public class Lion extends Mammals{
+	private static int id = 1;
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Lion.java]: can be duplicated in separate execution threads

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Lion.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+public class Lion extends Mammals{
+	private static int id = 1; 
+
+	/**
+	 * Parameterized Constructor
+	 * @param weight : Weight of the lion.
+	 * @param age : Age of the lion.
+	 */
+	public Lion(double weight, int age){
+		super("Lion" + (id++), weight, age, "Roar", 4, false, AnimalsAvailable.LION);
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Lion.java]: Hardcoded values should be avoided

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Elephant.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package ZooManagementSystem;
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Elephant.java]: refer to Lion class for suggested changes

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Reptiles.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
+package ZooManagementSystem;
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Reptiles.java]: refer to Mammals class for suggested changes

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Snake.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package ZooManagementSystem;
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Snake.java]: refer Lion class for suggested changes

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Crocodile.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package ZooManagementSystem;
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Crocodile.java]: refer Lion class for suggested changes

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Birds.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	 * @param flyHeight : The maximum height the bird can fly upto.
+	 * @param animalType : The type to which the bird belongs to.
+	 */
+	public Birds(String name, double weight, int age, String sound,
+			int numberOfLegs, boolean canFly, double flyHeight, AnimalsAvailable animalType){
+
+		super(name, weight, age, sound, numberOfLegs, AnimalTypes.BIRD, animalType);
+
+		this.canFly = canFly;
+		this.flyHeight = flyHeight;
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Birds.java]: should be populated only if the bird can fly

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Birds.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
+package ZooManagementSystem;
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Birds.java]: refer to Mammals class for suggested changes

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Parrot.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
+package ZooManagementSystem;
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Parrot.java]: refer to Lion class for suggested changes

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Peacock.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package ZooManagementSystem;
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Peacock.java]: refer to Lion class for suggested changes

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zoo.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+	/**
+	 * Method to add the zone into the zoo.
+	 * @param zoneType : The type of the zone which is to be added.
+	 * @param capacity : The capacity of the zone i.e. how many cages it can have.
+	 * @param hasPark : Whether the zone has park or not.
+	 * @param hasCanteen : Whether the zone has canteen or not.
+	 * @return : True if the zone is added, false otherwise.
+	 */
+	public boolean addZone(AnimalTypes zoneType, int capacity, boolean hasPark, boolean hasCanteen){
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zoo.java]: A variable of dataType AnimalTypes is named zoneType. Does that seem right?

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zoo.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
+package ZooManagementSystem;
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zoo.java]: Use of name capacity for both zoo capacity and cage capacity is a bit confusing, the names could have been segregated like cageCapacity and zooCapacity

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zones.java [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+	/**
+	 * Method to remove the animal from the cage.
+	 * @param animal : The animal which is to be removed.
+	 * @return : True if the animal is removed, false otherwise.
+	 */
+	public boolean removeAnimal(AnimalsAvailable animal){
+		boolean removed = false;
+		for(Cages cage: listOfCages){
+			if(cage.getAnimalType().equals(animal)){
+				removed = cage.removeAnimal();
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zones.java]: call remove only if cage has animals

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zoo.java [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+			break;
+		}
+		for(Zones zone: listOfZone){
+			if(zone.getZoneType().equals(category) && zone.removeAnimal(animal)){
+				flag = true;
+				break;
+			}
+		}
+		if(!flag){
+			throw new AssertionError("Animal not found!");
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zoo.java]: avoid using assertions in code

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zoo.java [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+		}
+		return flag;
+	}
+
+	/**
+	 * Method to add animal to the cage.
+	 * @param animal : The animal which is to be added to the cage.
+	 * @return : True if the animal is successfully added to the cage.
+	 */
+	public boolean addAnimal(Animals animal){
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zoo.java]: how are we ensuring that the zone or cage can hold no more animals without checking capacity?

File: Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zones.java [hunk 17]
```diff
@@ -0,0 +1,138 @@
+package ZooManagementSystem;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Niraj Gupta
+ * Class to represent the zones which are available in the zoo.
+ */
+public class Zones{
```
Comment (meta-simran-kapoor) [inline Software-Construction/Inheritance-And-Composition/ZooManagementSystem/Zones.java]: add zone id

**Review by ghost — APPROVED** [review]
Please keep suggested changes in mind

**Pull request merged**